more_itertools/more.py

```diff
@@ -129,24 +140,57 @@ def peek(self, default=_marker):
         provided, raise ``StopIteration``.
 
         """
-        if not hasattr(self, '_peek'):
+        if not self._cache:
             try:
-                self._peek = next(self._it)
+                self._cache.append(next(self._it))
             except StopIteration:
                 if default is _marker:
                     raise
                 return default
-        return self._peek
+        return self._cache[0]
 
     def __next__(self):
         ret = self.peek()
-        del self._peek
+        self._cache.popleft()
         return ret
 
     def next(self):
         # For Python 2 compatibility
         return self.__next__()
 
+    def _get_slice(self, index):
+        start = index.start
+        stop = index.stop
+
+        if (
+            ((start is not None) and (start < 0)) or
+            ((stop is not None) and (stop < 0))
+        ):
+            raise ValueError('Negative indexing not supported')
+
+        cache_len = len(self._cache)
+
+        if stop is None:
+            while True:
+                try:
+                    self._cache.append(next(self._it))
```

Comment:
I wonder if we could do this with .extend and islice() and save the overhead of spinning around in the while loop.